### PR TITLE
Re-use original source file imports for efficient incremental type-checking

### PR DESCRIPTION
**Pull request by devversion**
See individual commits. The new import manager is more complex, but basically it will allow us to:

- simplify the consumer side of the manager to generate imports + inserting them into TS transforms/contents
- improve incremental re-use with type checking— by allowing us to avoid structural source file changes
- share the same logic for schematics where we are solving similar problems— intending to re-use existing imports

This PR also adds tests for import generation. Something we don't have right now.
In a follow-up we can replace/remove the schematics import manager then. 

Related to: https://github.com/angular/angular/pull/53521#pullrequestreview-1778130879

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    directReuseCache: new Map(),
+    namespaceImportReuseCache: new Map(),
+  };
+
+  constructor(private _config: Partial<ImportManagerConfig> = {}) {}
+
+  /** Generates a side-effect import for the given module. */
+  generateSideEffectImport(requestedFile: ts.SourceFile, moduleSpecifier: string) {
+    if (!this.newImports.has(requestedFile)) {
+      this.newImports.set(requestedFile, {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:94]: Good point. Done!

File: packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+  };
+  private reuseGeneratedImportsTracker: ReuseGeneratedImportsTracker = {
+    directReuseCache: new Map(),
+    namespaceImportReuseCache: new Map(),
+  };
+
+  constructor(private _config: Partial<ImportManagerConfig> = {}) {}
+
+  /** Generates a side-effect import for the given module. */
+  generateSideEffectImport(requestedFile: ts.SourceFile, moduleSpecifier: string) {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:92]: Done!

File: packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    this.newImports.get(requestedFile)!.sideEffectImports.add(moduleSpecifier as ModuleName);
+  }
+
+  /**
+   * Adds an import to the given source-file and returns a TypeScript
+   * expression that can be used to access the newly imported symbol.
+   */
+  addImportToSourceFile(request: ImportRequest<ts.SourceFile>&
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_manager.ts:108]: Done!

File: packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_typescript_transform.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+      sourceFile = ts.visitEachChild(sourceFile, nodeVisitor, ctx);
+
+      // Filter out the existing imports and the source file body. All new statements
+      // will be inserted between them.
+      const existingImports = sourceFile.statements.filter(stmt => isImportStatement(stmt));
+      const body = sourceFile.statements.filter(stmt => !isImportStatement(stmt));
+      const extraStatements = extraStatementsForFiles?.get(sourceFile.fileName) ?? [];
+
+      return ctx.factory.updateSourceFile(
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_typescript_transform.ts:67]: Yeah, this is the same as before. The utility `addImports` used by the Ivy transforms also transformed the `sourceFile` using `updateSourceFile` — that should be fine/identical

File: packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_typescript_transform.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      if (!affectedFiles.has(sourceFile.fileName) &&
+          !extraStatementsForFiles?.has(sourceFile.fileName)) {
+        return sourceFile;
+      }
+
+      sourceFile = ts.visitEachChild(sourceFile, nodeVisitor, ctx);
+
+      // Filter out the existing imports and the source file body. All new statements
+      // will be inserted between them.
+      const existingImports = sourceFile.statements.filter(stmt => isImportStatement(stmt));
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/translator/src/import_manager/import_typescript_transform.ts:63]: Took this from `addImports`. Improved!

**Review by crisbeto — APPROVED** [review]

**Comment by devversion** [comment]
[Passing TGP](https://fusion2.corp.google.com/presubmit/616060787/OCL:616060787:BASE:616113992:1710510744714:32f2cd9d;groups=NewlyFailing,PossiblyNewlyFailing/targets). I've landed all preparation changes (making `magic_string` usable in g3)

**Comment by alxhub** [comment]
This PR was merged into the repository by commit d01576b104fb580910a49ec0deb91fb6bb220b5b.

**Pull request closed without merging**